selectors/selectors.md

```diff
@@ -302,67 +265,152 @@ If it's a file in some directory, you can also start at a deeper level:
-	| SelectRange "r"
-	| SelectRecursive "A"
-	| SelectUnion "u"
+## SelectorEnvelope is the recommended top-level value for serialized messages
+## that don't have established existing context with marks the start of a selector:
+## it's a single-member union used to kick us towards "nominative typing".
+##
+## See https://github.com/ipld/go-ipld-prime/blob/0692e3b8cd7f231fe5d9d16a103bbbacb23dbdb5/doc/schema.md#using-schema-match-checking-as-version-detection
+## for a background on the theory behind this gentle-nominative concept.
+type SelectorEnvelope union {
+	| Selector "selector"
 } representation keyed
 
 type Selector union {
-	| SelectorComplex map
-	| SelectTrue bool
-} representation kinded
-
-# SelectTrue is a terminal selector which will not explore any more fields,
-# and marks the node it's applied to as a match.
-type SelectTrue bool
+	| Matcher "."
+	| ExploreAll "a"
+	| ExploreFields "f" # note "ExplorePath" is a degenerate case of ExploreFields.
+	| ExploreIndex "i"
+	| ExploreRange "r"
+	| ExploreRecursive "R"
+	| ExploreUnion "|"
+	| ExploreConditional "&"
+	| ExploreRecursiveEdge "@" # sentinel value; only valid in some positions.
+} representation keyed
 
-# SelectAll considers all members of the data; think of it as similar to `./*`.
-type SelectAll struct {
-	next Selector (alias ":")
+## ExploreAll is similar to a `*` -- it traverses all elements of an array,
+## or all entries in a map, and applies a next selector to the reached nodes.
+type ExploreAll struct {
+	next Selector (alias ">")
 }
 
-# SelectFields will explore each of the fields listed, and yields a distinct
-# selector for continuing to explore and match on each reached node.
-type SelectFields map {String:Selector}
+## ExploreFields traverses named fields in a map (or equivalently, struct, if
+## traversing on typed/schema nodes) and applies a next selector to the
+## reached nodes.
+##
+## Note that a concept of "ExplorePath" (e.g. "foo/bar/baz") can be represented
+## as a set of three nexted ExploreFields selectors, each specifying one field.
+## (For this reason, we don't have a special "ExplorePath" feature; use this.)
+type ExploreFields struct {
+	fields {String:Selector} (alias "f>")
+}
 
-# SelectIndex considers a specific item in a list by its offset.
-type SelectIndex struct {
+## ExploreIndex traverses a specific index in a list, and applies a next
+## selector to the reached node.
+type ExploreIndex struct {
 	index Int (alias "i")
-	next Selector (alias ":")
+	next Selector (alias ">")
 }
 
-# SelectRange considers each of the items in a range of positions in a list.
-type SelectRange struct {
+## ExploreIndex traverses a list, and for each element in the range specified,
+## will apply a next selector to those reached nodes.
+type ExploreRange struct {
 	start Int (alias "^")
 	end Int (alias "$")
-	next Selector (alias ":")
+	next Selector (alias ">")
+}
+
+## ExploreRecursive traverses some structure using fragments of another
```

Comment:
is it considered value to have an ExploreRecursive inside another ExploreRecursive, and if so, is it reasonable to assume any ExploreRecursiveEdges reference their immediate parent?

Reply:
Yes and yes, I think.

I had some earlier draft where nested recursion was forbidden, but then a use case was proposed which actually needed it, so I dropped that text :(  So, the first yes.

I couldn't think of any cases where it's sane for ExploreRecursiveEdge to refer to anything other than their immediate parent, though.  We're not trying to build gotos here; that'd be... bad.  So, the second yes!

Reply:
One other moderately terrifying thing I think should probably be supported here: there can be two (or more) different ExploreRecursiveEdge both referring back to the same ExploreRecursive.  Roughly analogous to the use of a 'continue' statement in a for-loop.

Reply:
Pushed some additional docs here to clarify and expand upon this. :+1: